homeassistant/components/media_player/directv.py

```diff
@@ -47,105 +54,238 @@
+        await self._async_get_dtv_instance()
 
-    def update(self):
+    async def _async_discover_directv_client_devices(self, now=None):
+        """Discover new client devices connected to the main DVR."""
+        known_devices = self.hass.data.get(DATA_DIRECTV, set())
+        discovered_devices = []
+        dtvs = []
+
+        # Attempt to discover additional RVU units
+        if now:
+            _LOGGER.debug("%s: Scheduled discovery of DirecTV devices on %s",
+                          self.entity_id, self._host)
+        else:
+            _LOGGER.debug("%s: Initial discovery of DirecTV devices on %s",
+                          self.entity_id, self._host)
+
+        _LOGGER.debug("%s: Current known devices: %s",
+                      self.entity_id, known_devices)
+
+        try:
+            resp = await self._async_get_locations()
+        except requests.exceptions.RequestException as ex:
+            _LOGGER.debug("%s: Request exception %s trying to discover "
+                          "new clients", self.entity_id, ex)
+
+        for loc in resp.get('locations') or []:
+            if 'locationName' not in loc or 'clientAddr' not in loc or\
+               loc.get('clientAddr') == DEFAULT_DEVICE:
+                continue
+
+            # Make sure that this device is not already configured
+            # Comparing based on host (IP) and clientAddr.
+            if (self._host, loc['clientAddr']) in known_devices:
+                _LOGGER.debug("%s: Discovered device %s on host %s with "
+                              "client address %s is already "
+                              "configured",
+                              self.entity_id,
+                              str.title(loc['locationName']),
+                              self._host, loc['clientAddr'])
+            else:
+                _LOGGER.debug("%s: Adding discovered device %s with"
+                              " client address %s",
+                              self.entity_id,
+                              str.title(loc['locationName']),
+                              loc['clientAddr'])
+                discovered_devices.append({
+                    CONF_NAME: str.title(loc['locationName']),
+                    CONF_HOST: self._host,
+                    CONF_PORT: self._port,
+                    CONF_DEVICE: loc['clientAddr'],
+                    CONF_DISCOVER_CLIENTS: False,
+                    'entities': self._async_add_entities
+                })
+
+        if discovered_devices:
+            _LOGGER.debug("%s: Adding %s new DirecTV entities to HASS",
+                          self.entity_id, len(discovered_devices))
+
+            for new_device in discovered_devices:
+                dtvs.append(DirecTvDevice(**new_device))
+                self.hass.data.setdefault(DATA_DIRECTV, set()).add(
+                    (new_device[CONF_HOST], new_device[CONF_DEVICE]))
+
+            self._async_add_entities(dtvs)
```

Comment:
We shouldn't add new entities from within another entity. That's really weird. If we need to add entities after platform setup, make another instance that handles this that is not an entity.